Drag and drop

marco9446 opened the pull request:
Added the possibility to drag and drop an element from the right nav bar into the slide canvas.
All the imported element are lazy loaded and injected in the slide area.
The drop target is now hard coded to the rood of slide area (paper-element) but it already prepared to work with a customisable target

Commit:
implemented electron and modified structure

Commit:
modified title and removed console.log

Commit:
implemented some basic drag and drop capabilities

now a use is able to drag an element from the available question panel
(should work with all the question type as long as is used the same
name of the bower components) and drop it in the canvas

Commit:
automatically understand if the user is using electron

Commit:
improved highlight drop target function

Commit:
improved import and inject function

triglian commented:
### Tests

When testing an element (unit-test) you should minimally test
1. Public methods
2. setting and getting public properties
3. content nodes (check if content is distributed like you want it to)
4. render (if sub elements are where they need to be, if dom-if templates work as they should etc.)

When testing a feature normally there's more than a few elements involved. There you need to programmatically perform a user interaction (e.g drag and drop) and check if everything is at it's place. Bear in mind that how an element renders is also important

Check the designer for tests

Commit:
prefixed element with qea- and removed scripts directory

Commit:
Add polylint to the building process and updated the readme

Commit:
removed slide references  and made the whole stage droppable

Commit:
style changed for toolbar and list

Also added the possibility to group questions

marco9446 commented:
I changed the name of components in order to follow convention and avoid name collision.
I did't find a linter for javascript, compatible with polymer and gulp, so i just implemented polylint.
The droppable are is now the whole stage and i made some changes in the style to remain as neutral as possible and you have also the ability to create group in the draggable list.
I'll make some test after having converted the project to ES6

Commit:
translate elements in ES6 and built a basic transpiler  for browser  compatibility

Commit:
Merge branch 'es6-conversion' into drag-and-drop

Commit:
fixed transpiler and added a linter for javascript

Commit:
Added property and style editor (just UI)

Commit:
create a element "library" to use hydrolysis and add behavior

Commit:
added documentation and the ability to change property

marco9446 commented:
I have decided to abandon the idea of using a custom element to wrap selected item since i haven't found any good reason to proceed in that direction.
Right now I just wrap the dynamically created elements in a DIV

Commit:
create a basic test for question-type

Commit:
add lending page and momentaneously remove some features

Commit:
Merge branch 'test' into drag-and-drop

Commit:
Merge branch test

Commit:
fix tests names

Commit:
add test for property analyser

Commit:
fix bug property analyzer and add test on it

Commit:
add test for qea-question-list

triglian committed:
minor corrections

triglian committed:
add instructions for wct

triglian committed:
add gulp-polylint to deps

triglian committed:
type var name correctly

triglian commented on app/elements/qea-property-analyzer/qea-property-analyzer.html:
```diff
@@ -0,0 +1,138 @@
+      }
+
+      // return the list of behaviors of the element analyzed
+      getBehaviors(analyzer){
```
You have a getter for behaviours and then you have the `getBehaviors` function. What's their difference? As an interface, isn't this confusing?

marco9446 replied:
The getter for behaviours is a polymer notation to inject behaviours inside that component instead the getBehaviors function return the behaviours use by the element analysed by hydrolysis .

triglian replied:
then maybe `getBehaviors` -> `getElementBehaviors`?

triglian commented on app/elements/qea-property-analyzer/qea-property-analyzer.html:
```diff
@@ -0,0 +1,138 @@
+  <script>
+  (function () {
+    'use strict'
+
```
I assume all the properties and methods that don't start with an `_` are supposed to be accessible publicly right?

Make sure this is the case for all files

marco9446 replied:
fixed

triglian commented on app/elements/qea-question-type/qea-question-type.html:
```diff
@@ -0,0 +1,130 @@
+      }
+
+      _currentlyDragging(dragDetail){
+         // console.log(dragDetail);
```
remove the loggint statement.

triglian commented on app/test/qea-property-analyzer-test.html:
```diff
@@ -0,0 +1,81 @@
+        });
+
+
+        test('test scanElemen method', function(done) {
```
Your don't need "check" or  "test" on your test messages, it is implied.  You should try to thing of the message like :
(test that the) "scanElement method succeeds for asq elements'".  Check here for some examples: https://github.com/Polymer/designer/blob/master/test/path/path_test.html. Please change all affected messages across all test files.

triglian commented on electron.js:
```diff
@@ -0,0 +1,39 @@
+'use strict';
```
If you borrowed this from somewhere, it is good if you mention it on the top of the file.

marco9446 replied:
I took it from the [Quick Start guide](http://electron.atom.io/docs/latest/tutorial/quick-start/) of electron so i don't think it's necessary to mention

Commit:
removed log and fixed test names

Commit:
fixed private properties

Commit:
fixed others private properties notations

triglian commented on bower.json:
```diff
@@ -1,5 +1,6 @@
 {
   "name": "editor-project",
```
the name is `question-editor-asq`

triglian commented on app/test/qea-property-analyzer-test.html:
```diff
@@ -0,0 +1,81 @@
+        });
+
+
+        test('retrun an analyzer object from scanElement method', function(done) {
```
`retrun`

Commit:
fixed spelling mistake and name

Commit:
fixed license and added repository path

triglian committed:
slight modifications to test readme

triglian commented:
There should be a command that runs the test when you do `npm test`

Commit:
added test command and fixed qea-property-analyzer test

Commit:
modified test command
